src/serac/physics/utilities/functional/functional.hpp

```diff
@@ -357,20 +388,246 @@ class Functional<test(trial), execution_policy> : public mfem::Operator {
    */
   class Gradient : public mfem::Operator {
+
+    template < typename T >
+    struct Array3D{
+      Array3D() = default;
+      Array3D(int n1, int n2, int n3) : strides{n2 * n3, n3, 1}, data(n1 * n2 * n3) {}
+      auto & operator()(int i, int j, int k) { return data[i * strides[0] + j * strides[1] + k * strides[2]]; }
+      int strides[3];
+      std::vector < T > data;
+    };
+
   public:
     /**
      * @brief Constructs a Gradient wrapper that references a parent @p Functional
      * @param[in] f The @p Functional to use for gradient calculations
      */
-    Gradient(Functional& f) : mfem::Operator(f.Height()), form(f){};
+    Gradient(Functional<test(trial)> & f) : mfem::Operator(f.Height(), f.Width()), form(f), sparsity_pattern_initialized(false) {};
 
     virtual void Mult(const mfem::Vector& x, mfem::Vector& y) const override { form.GradientMult(x, y); }
+    
+    mfem::Vector operator()(const mfem::Vector& x) const { 
+      mfem::Vector y(form.Height());
+      form.GradientMult(x, y); 
+      return y;
+    }
+
+    void initialize_sparsity_pattern() {
+
+      mfem::Array<int> test_dofs;
+      mfem::Array<int> trial_dofs;
+
+      int test_vdim  = form.test_space_->GetVDim();
+      int trial_vdim = form.trial_space_->GetVDim();
+
+      form.test_space_->GetElementDofs(0, test_dofs);
+      form.trial_space_->GetElementDofs(0, trial_dofs);
+      int num_elements = form.test_space_->GetNE();
+      int dofs_per_test_element = test_dofs.Size();
+      int dofs_per_trial_element = trial_dofs.Size();
+      int entries_per_element = dofs_per_test_element * dofs_per_trial_element;
+
+      form.test_space_->GetBdrElementDofs(0, test_dofs);
+      form.trial_space_->GetBdrElementDofs(0, trial_dofs);
+      int num_boundary_elements = form.test_space_->GetNBE();
+      int dofs_per_test_boundary_element = test_dofs.Size();
+      int dofs_per_trial_boundary_element = trial_dofs.Size();
+      int entries_per_boundary_element = dofs_per_test_boundary_element * test_vdim * dofs_per_trial_boundary_element * trial_vdim;
+
+      int num_infos[2] = {
+        (form.domain_integrals_.size() > 0) * entries_per_element * num_elements,
+        (form.boundary_integrals_.size() > 0) * entries_per_boundary_element * num_boundary_elements
+      };
+
+      std::vector < ::detail::elem_info > infos;
+      infos.reserve(num_infos[0] + num_infos[1]);
+
+      if (form.domain_integrals_.size() > 0) {
+        bool on_boundary = false;
+
+        for (int e = 0; e < num_elements; e++) {
+          form.test_space_->GetElementDofs(e, test_dofs);
+          form.trial_space_->GetElementDofs(e, trial_dofs);
+
+          const mfem::Array<int> & test_native_to_lexicographic = dynamic_cast<const mfem::TensorBasisElement *>(form.test_space_->GetFE(0))->GetDofMap();
+          const mfem::Array<int> & trial_native_to_lexicographic = dynamic_cast<const mfem::TensorBasisElement *>(form.trial_space_->GetFE(0))->GetDofMap();
```

Comment:
We should either check to see if the `dynamic_cast` was successful or just use a `static_cast` here

Reply:
I believe we should be okay, since we're actually checking that all of the elements are of the appropriate type in `AddDomainIntegral`, `AddBoundaryIntegral`, but it's probably worth doing. It's also not clear to me whether or not GetDofMap actually depends on the instance or not. If this permutation is always the same for each element type, we should probably move it out of the loop.